File: windows/CuotaX/UpdateChecker.cs
// SPDX-License-Identifier: MIT

using System.Reflection;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CuotaX;

internal sealed class UpdateChecker : IDisposable
{
    private const string ApiUrl = "https://api.github.com/repos/jpnurmi/cuotax/compare";

    private readonly HttpClient _client;
    private readonly string? _buildCommit;

    internal UpdateChecker(HttpClient? client = null, string? buildCommit = null)
    {
        _client = client ?? new HttpClient { Timeout = TimeSpan.FromSeconds(15) };
        _buildCommit = buildCommit ?? Assembly.GetExecutingAssembly()
            .GetCustomAttributes<AssemblyMetadataAttribute>()
            .FirstOrDefault(attribute => attribute.Key == "BuildCommit")
            ?.Value;
    }

    internal async Task<bool> IsUpdateAvailableAsync(CancellationToken cancellationToken = default)
    {
        if (!IsCommit(_buildCommit))
        {
            return false;
        }

        using var request = new HttpRequestMessage(
            HttpMethod.Get,
            $"{ApiUrl}/{_buildCommit}...main"
        );
        request.Headers.Add("Accept", "application/vnd.github+json");
        request.Headers.Add("User-Agent", "CuotaX");
        using var response = await _client.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            return false;
        }

        return HasUpdate(await response.Content.ReadAsStringAsync(cancellationToken));
    }

    internal static bool HasUpdate(string json)
    {
        try
        {
            return JsonSerializer.Deserialize<Comparison>(json)?.AheadBy > 0;
        }
        catch (JsonException)
        {
            return false;
        }
    }

    public void Dispose() => _client.Dispose();

    private static bool IsCommit(string? value) =>
        value is { Length: 40 }
        && value.All(character => character is >= '0' and <= '9' or >= 'a' and <= 'f');

    private sealed record Comparison([property: JsonPropertyName("ahead_by")] int AheadBy);
}

File: macos/Sources/CuotaX/UpdateChecker.swift
// SPDX-License-Identifier: MIT

import Foundation

struct UpdateChecker: Sendable {
  typealias Loader = @Sendable (URLRequest) async throws -> Data

  private static let apiURL = "https://api.github.com/repos/jpnurmi/cuotax/compare"
  private let buildCommit: String?
  private let load: Loader

  init(
    buildCommit: String? = Bundle.main.object(forInfoDictionaryKey: "CuotaXBuildCommit") as? String,
    load: @escaping Loader = UpdateChecker.load
  ) {
    self.buildCommit = buildCommit
    self.load = load
  }

  func isUpdateAvailable() async -> Bool {
    guard let buildCommit, Self.isCommit(buildCommit),
      let url = URL(string: "\(Self.apiURL)/\(buildCommit)...main")
    else { return false }

    var request = URLRequest(url: url)
    request.timeoutInterval = 15
    request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
    request.setValue("CuotaX", forHTTPHeaderField: "User-Agent")
    do {
      return try Self.hasUpdate(in: await load(request))
    } catch {
      return false
    }
  }

  static func hasUpdate(in data: Data) throws -> Bool {
    try JSONDecoder().decode(Comparison.self, from: data).aheadBy > 0
  }

  private static func isCommit(_ value: String) -> Bool {
    value.count == 40 && value.allSatisfy { $0.isHexDigit && !$0.isUppercase }
  }

  private static func load(_ request: URLRequest) async throws -> Data {
    let (data, response) = try await URLSession.shared.data(for: request)
    guard let response = response as? HTTPURLResponse, (200..<300).contains(response.statusCode)
    else { throw URLError(.badServerResponse) }
    return data
  }
}

private struct Comparison: Decodable {
  let aheadBy: Int

  enum CodingKeys: String, CodingKey {
    case aheadBy = "ahead_by"
  }
}
